### ISSUE TITLE
where is lineart_standard (from white bg & black line) in comfyui_controlnet_aux

### ISSUE THREAD

**Issue by jakeute** [Fannovel16/comfyui_controlnet_aux#159]
<img width="773" alt="image" src="https://github.com/Fannovel16/comfyui_controlnet_aux/assets/47204876/685dd576-8ff0-4bf8-82ef-30370b216338">
I queried all the preprocessors, but unfortunately I couldn't find lineart_standard (from white bg & black line). Is it because the plug-in has not implemented it yet?

**Comment by jakeute** [comment]
I took a look at the code of the sd webui and found that its implementation was simple, and I migrated to comfyui manually

**Comment by jakeute** [comment]
class LineArtStandardPreprocessor:
    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "image": ("IMAGE",)
            },
            "optional": {
                "resolution": ("INT", {"default": 512, "min": 64, "max": 2048, "step": 64})
            }
        }

    RETURN_TYPES = ("IMAGE",)
    FUNCTION = "execute"
    CATEGORY = "comic"

    def execute(self, image, resolution=512, ):
        out_list = []
        for img in image:
            np_image = np.asarray(img * 255., dtype=np.uint8)
            img, remove_pad = resize_image_with_pad(np_image, resolution)
            x = img.astype(np.float32)
            g = cv2.GaussianBlur(x, (0, 0), 6.0)
            intensity = np.min(g - x, axis=2).clip(0, 255)
            intensity /= max(16, np.median(intensity[intensity > 8]))
            intensity *= 127
            result = intensity.clip(0, 255).astype(np.uint8)
            out_list.append(torch.from_numpy(remove_pad(result).astype(np.float32) / 255.0))
        return (torch.stack(out_list, dim=0),)

**Comment by jakeute** [comment]
```
class LineArtStandardPreprocessor:
    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "image": ("IMAGE",)
            },
            "optional": {
                "resolution": ("INT", {"default": 512, "min": 64, "max": 2048, "step": 64})
            }
        }

    RETURN_TYPES = ("IMAGE",)
    FUNCTION = "execute"
    CATEGORY = "comic"

    def execute(self, image, resolution=512, ):
        out_list = []
        for img in image:
            np_image = np.asarray(img * 255., dtype=np.uint8)
            img, remove_pad = resize_image_with_pad(np_image, resolution)
            x = img.astype(np.float32)
            g = cv2.GaussianBlur(x, (0, 0), 6.0)
            intensity = np.min(g - x, axis=2).clip(0, 255)
            intensity /= max(16, np.median(intensity[intensity > 8]))
            intensity *= 127
            result = intensity.clip(0, 255).astype(np.uint8)
            out_list.append(torch.from_numpy(remove_pad(result).astype(np.float32) / 255.0))
        return (torch.stack(out_list, dim=0),)
```

**Comment by jakeute** [comment]
```
def resize_image_with_pad(input_image, resolution, skip_hwc3=False):
    if skip_hwc3:
        img = input_image
    else:
        img = HWC3(input_image)
    H_raw, W_raw, _ = img.shape
    k = float(resolution) / float(min(H_raw, W_raw))
    interpolation = cv2.INTER_CUBIC if k > 1 else cv2.INTER_AREA
    H_target = int(np.round(float(H_raw) * k))
    W_target = int(np.round(float(W_raw) * k))
    img = cv2.resize(img, (W_target, H_target), interpolation=interpolation)
    H_pad, W_pad = pad64(H_target), pad64(W_target)
    img_padded = np.pad(img, [[0, H_pad], [0, W_pad], [0, 0]], mode='edge')

    def remove_pad(x):
        return safer_memory(x[:H_target, :W_target])

    return safer_memory(img_padded), remove_pad

def pad64(x):
    return int(np.ceil(float(x) / 64.0) * 64 - x)

def safer_memory(x):
    # Fix many MAC/AMD problems
    return np.ascontiguousarray(x.copy()).copy()
```

**Comment by jakeute** [comment]
<img width="1026" alt="image" src="https://github.com/Fannovel16/comfyui_controlnet_aux/assets/47204876/bab48a7c-7964-45c8-9cc4-e710cc571ad9">

**Comment by apwep** [comment]
You may try the author's comfy_controlnet_preprocessors plugin below:
https://github.com/Fannovel16/comfy_controlnet_preprocessors

**Comment by Fannovel16** [comment]
Added in https://github.com/Fannovel16/comfyui_controlnet_aux/commit/432591791e66d8bb6d4800d62b5d69e937e07108
